From b21d412cb19701bef2584df111b644bac1afdccf Mon Sep 17 00:00:00 2001
From: Brad Ho <Brad_Ho@phoenix.com>
Date: Tue, 9 Nov 2021 01:08:29 -0800
Subject: [PATCH] Improve phoenix send trap function 

Note:
1. Change return type to void that will case core dump happened     
2. Change ip value to  default set if not provide address
3. Change to use pointer value to response previous_alert     
4. Remove use strdup() function, direct to use
   const_cast<char*> for c_str() avoid core dump problem.

---
 snmp.hpp              |  2 +-
 snmp_notification.cpp | 33 +++++++++++++++++++++++----------
 snmp_notification.hpp | 26 ++++++++++++++------------
 3 files changed, 38 insertions(+), 23 deletions(-)

diff --git a/snmp.hpp b/snmp.hpp
index bd5b752..ae9a840 100644
--- a/snmp.hpp
+++ b/snmp.hpp
@@ -25,7 +25,7 @@ template <typename T, typename... ArgTypes> void sendTrap(ArgTypes&&... tArgs)
  * @param[in] tArgs - arguments for the trap.
  */
 
-template <typename T, typename... ArgTypes> bool phoenixSendTrap(ArgTypes&&... tArgs)
+template <typename T, typename... ArgTypes> void phoenixSendTrap(ArgTypes&&... tArgs)
 {
     T obj(std::forward<ArgTypes>(tArgs)...);
     obj.phoenixSendTrap();
diff --git a/snmp_notification.cpp b/snmp_notification.cpp
index 40307b3..d29d1e7 100644
--- a/snmp_notification.cpp
+++ b/snmp_notification.cpp
@@ -78,6 +78,10 @@ std::string getDestinationAddresses(uint8_t set)
         ipaddr = addressList[set][ipaddrtStr];
         log<level::INFO>("Use destination address from config.",
             entry("DEST_ADDR=%s", ipaddr.c_str()));
+    } else {
+        ipaddr = "192.168.122.1";
+        log<level::INFO>("Use destination address from default value.",
+            entry("DEST_ADDR=%s", ipaddr.c_str()));
     }
     return ipaddr;
 }
@@ -204,7 +208,7 @@ void Notification::sendTrap()
     }
 }
 
-bool Notification::phoenixSendTrap()
+void Notification::phoenixSendTrap()
 {
     std::string comm = phoenix::getCommunityString();
     std::string ip;
@@ -221,15 +225,17 @@ bool Notification::phoenixSendTrap()
     session.callback = nullptr;
     session.callback_magic = nullptr;
 
-    if( getDest() > (phoenix::getNumberOfDestinationsSize() - 1) )
+    if( getDest() > (phoenix::getNumberOfDestinationsSize() - 1) ||
+        getDest() <= 0)
     {
         log<level::ERR>("SNMP unable to use destination",
                             entry("DESTINATION=%d", getDest()));
-        return false;
+        setFlag(false);
+        return;
     }
 
     ip = phoenix::getDestinationAddresses(getDest()) + ":162";
-    session.peername = strdup(ip.c_str());
+    session.peername = const_cast<char*>(ip.c_str());
     // create the session
     auto ss = snmp_add(
         &session,
@@ -239,7 +245,8 @@ bool Notification::phoenixSendTrap()
     {
         log<level::ERR>("Unable to get the snmp session.");
         elog<InternalFailure>();
-        return false;
+        setFlag(false);
+        return;
     }
 
     // Wrap the raw pointer in RAII
@@ -252,7 +259,8 @@ bool Notification::phoenixSendTrap()
     {
         log<level::ERR>("Failed to create notification PDU");
         elog<InternalFailure>();
-        return false;
+        setFlag(false);
+        return;
     }
 
     pdu->trap_type = SNMP_TRAP_ENTERPRISESPECIFIC;
@@ -267,7 +275,8 @@ bool Notification::phoenixSendTrap()
         log<level::ERR>("Failed to add the SNMP var(trapID)");
         snmp_free_pdu(pdu);
         elog<InternalFailure>();
-        return false;
+        setFlag(false);
+        return;
     }
 
     auto objectList = getFieldOIDList();
@@ -280,7 +289,8 @@ bool Notification::phoenixSendTrap()
             log<level::ERR>("Failed to add the SNMP var");
             snmp_free_pdu(pdu);
             elog<InternalFailure>();
-            return false;
+            setFlag(false);
+            return;
         }
     }
     // pdu is freed by snmp_send
@@ -288,12 +298,15 @@ bool Notification::phoenixSendTrap()
     {
         log<level::ERR>("Failed to send the snmp trap.");
         elog<InternalFailure>();
-        return false;
+        setFlag(false);
+        return;
     }
     log<level::DEBUG>("Sent SNMP Trap");
     log<level::INFO>("Send to specified SMTP address to selected destination",
                         entry("ADDR=%s", ip.c_str()));
-    return true;
+
+    /*Set previous alert flag*/
+    setFlag(true);
 }
 } // namespace snmp
 } // namespace network
diff --git a/snmp_notification.hpp b/snmp_notification.hpp
index d1c8faf..7ae6ad1 100644
--- a/snmp_notification.hpp
+++ b/snmp_notification.hpp
@@ -104,7 +104,7 @@ class Notification
      *          manager.
      */
     void sendTrap();
-    bool phoenixSendTrap();
+    void phoenixSendTrap();
 
   protected:
     /** @brief Add the variable in the snmp pdu object.
@@ -132,6 +132,7 @@ class Notification
     virtual std::vector<Object> getFieldOIDList() = 0;
 
     virtual int getDest() = 0;
+    virtual void setFlag(bool) = 0;
 };
 
 class TestErrorNotification;
@@ -151,6 +152,7 @@ class OBMCErrorNotification : public Notification
     int32_t OBMCErrorSeverity;
     std::string OBMCErrorMessage;
     int destTest = 0;
+    bool *flag;
 
   public:
     OBMCErrorNotification() = default;
@@ -224,6 +226,10 @@ class OBMCErrorNotification : public Notification
     {
         return destTest;
     }
+    void setFlag(bool status) override
+    {
+        *flag = status;
+    }
     friend class TestErrorNotification;
 };
 
@@ -257,8 +263,8 @@ class PhoenixErrorNotification : public Notification
     uint32_t languageCode;
     uint64_t mftID;
     uint32_t systemID;
-    std::string msg;
     int destTest;
+    bool *flag;
 
   public:
     PhoenixErrorNotification() = default;
@@ -276,13 +282,12 @@ class PhoenixErrorNotification : public Notification
      */
     PhoenixErrorNotification(uint32_t seqN, uint32_t type,
                           uint32_t eventType, uint32_t eventOffset,
-                          //uint32_t guid, uint64_t timeStamp,
                           uint32_t guid, uint32_t timeStamp,
                           uint32_t utc, uint32_t trapSourceType, uint32_t eventSourceType,
                           uint32_t eventSeverity, uint32_t sensorDevice, uint32_t sensorNumber,
                           uint32_t entity, uint32_t entityInstance, uint32_t eventData1,
                           uint32_t eventData2, uint32_t eventData3, uint32_t languageCode,
-                          uint64_t mftID, uint32_t systemID, std::string msg, int dest) :
+                          uint64_t mftID, uint32_t systemID, int dest, bool *flag) :
         seqNum(seqN), sensorType(type),
         eventType(eventType), eventOffset(eventOffset), guid(guid),
         timeStamp(timeStamp), utc(utc), trapSourceType(trapSourceType),
@@ -290,7 +295,7 @@ class PhoenixErrorNotification : public Notification
         sensorDevice(sensorDevice), sensorNumber(sensorNumber), entity(entity),
         entityInstance(entityInstance), eventData1(eventData1), eventData2(eventData2),
         eventData3(eventData3), languageCode(languageCode), mftID(mftID), systemID(systemID),
-        msg(msg), destTest(dest)
+        destTest(dest), flag(flag)
     {
     }
   protected:
@@ -444,13 +449,6 @@ class PhoenixErrorNotification : public Notification
         tuple = std::tuple<OID, OID_LEN, Type, Value>(id, idLen, type,
                                                       systemID);
 
-        objectList.emplace_back(std::move(tuple));
-
-        id = {1, 3, 6, 1, 4, 1, 49871, 1, 0, 1, 21};
-        type = getASNType<decltype(msg)>();
-        tuple = std::tuple<OID, OID_LEN, Type, Value>(id, idLen, type,
-                                                      msg);
-
         objectList.emplace_back(std::move(tuple));
         return objectList;
     }
@@ -459,6 +457,10 @@ class PhoenixErrorNotification : public Notification
     {
         return destTest;
     }
+    void setFlag(bool status) override
+    {
+        *flag = status;
+    }
     friend class TestErrorNotification;
 };
 } // namespace snmp
